server/src/main/java/org/elasticsearch/index/mapper/PostParsePhase.java

```diff
@@ -0,0 +1,533 @@
+            checkField(field);
+            List<Number> values = document.getFields().stream()
+                .filter(f -> Objects.equals(f.name(), field))
+                .filter(f -> f.fieldType().docValuesType() == DocValuesType.SORTED_NUMERIC)
+                .map(IndexableField::numericValue)
+                .sorted()
+                .collect(Collectors.toList());
+            return sortedNumericDocValues(values);
+        }
+
+        @Override
+        public SortedSetDocValues getSortedSetDocValues(String field) throws IOException {
+            List<BytesRef> values = document.getFields().stream()
+                .filter(f -> Objects.equals(f.name(), field))
+                .filter(f -> f.fieldType().docValuesType() == DocValuesType.SORTED_SET)
+                .map(IndexableField::binaryValue)
+                .sorted()
+                .collect(Collectors.toList());
+            return sortedSetDocValues(values);
+        }
+
+        @Override
+        public FieldInfos getFieldInfos() {
+            return new FieldInfos(new FieldInfo[0]);
+        }
+
+        @Override
+        public void document(int docID, StoredFieldVisitor visitor) throws IOException {
+            List<IndexableField> fields = document.getFields().stream()
+                .filter(f -> f.fieldType().stored())
+                .collect(Collectors.toList());
+            for (IndexableField field : fields) {
+                FieldInfo fieldInfo = fieldInfo(field.name());
+                if (visitor.needsField(fieldInfo) != StoredFieldVisitor.Status.YES) {
+                    continue;
+                }
+                if (field.numericValue() != null) {
+                    Number v = field.numericValue();
+                    if (v instanceof Integer) {
+                        visitor.intField(fieldInfo, v.intValue());
+                    } else if (v instanceof Long) {
+                        visitor.longField(fieldInfo, v.longValue());
+                    } else if (v instanceof Float) {
+                        visitor.floatField(fieldInfo, v.floatValue());
+                    } else if (v instanceof Double) {
+                        visitor.doubleField(fieldInfo, v.doubleValue());
+                    }
+                } else if (field.stringValue() != null) {
+                    visitor.stringField(fieldInfo, field.stringValue().getBytes(StandardCharsets.UTF_8));
+                } else if (field.binaryValue() != null) {
+                    // We can't just pass field.binaryValue().bytes here as there may be offset/length
+                    // considerations
+                    byte[] data = new byte[field.binaryValue().length];
+                    System.arraycopy(field.binaryValue().bytes, field.binaryValue().offset, data, 0, data.length);
+                    visitor.binaryField(fieldInfo, data);
+                }
+            }
+        }
```

Comment:
is this effectively a fork of memory index, or the same code?

Reply:
It's completely new, MemoryIndex does not support stored fields.

Reply:
ok! could we add documentation about what all these bits are for, and what this has in common and not in common with memory index? I can see folks trying to refactor this to use memory index in the future without exactly knowing what's going on